src/Altinn.Apps/AppTemplates/AspNet/Altinn.App.PlatformServices/Options/Altinn2Provider/Altinn2CodeListProvider.cs

```diff
@@ -0,0 +1,91 @@
+#nullable enable
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+using Altinn.App.Common.Models;
+using Altinn.App.PlatformServices.Options;
+
+namespace Altinn.App.PlatformServices.Options.Altinn2Provider
+{
+    /// <summary>
+    /// Implementation of a IAppOptionsProviders for the old altinn2 apis
+    /// </summary>
+    public class Altinn2CodeListProvider : IAppOptionsProvider
+    {
+        /// <summary>
+        /// Mapping function to get from the altinn2 model to altinn 3 option
+        /// </summary>
+        private readonly Func<MetadataCodeListCodes, AppOption> _transform;
+
+        /// <summary>
+        /// Filter function in case you only want a subset of the altinn2 codelist
+        /// </summary>
+        private readonly Func<MetadataCodeListCodes, bool>? _filter;
+
+        /// <summary>
+        /// id for use in altinn2 api
+        /// </summary>
+        private readonly string _metadataApiId;
+
+        /// <summary>
+        /// version of the code list in the altinn2 metadata api
+        /// </summary>
+        private readonly int? _codeListVersion;
+
+        /// <summary>
+        /// Cache for options as altinn2 options are static
+        /// </summary>
+        private readonly Dictionary<string, AppOptions> _cachedOptions = new();
+
+        /// <inheritdoc />
+        public string Id { get;  private set; }
+
+        /// <summary>
+        /// <see cref="Altinn.App.PlatformServices.Options.Altinn2Provider.Altinn2CodeListOptionsBuilder.Add(string, Func{MetadataCodeListCodes, AppOption}, Func{MetadataCodeListCodes, bool}?, string?, int?)" />
+        /// </summary>
+        public Altinn2CodeListProvider(string id, Func<MetadataCodeListCodes, AppOption> transform, Func<MetadataCodeListCodes, bool>? filter, string? metadataApiId = null, int? codeListVersion = null)
+        {
+            Id = id; // id in layout definitions
+            _metadataApiId = metadataApiId ?? id; // codelist id in api (often the same as id, but if the same codelist is used with different filters, it has to be different)
+            _transform = transform;
+            _filter = filter;
+            _codeListVersion = codeListVersion;
+        }
+
+        /// <inheritdoc/>
+        public async Task<AppOptions> GetAppOptionsAsync(string language, Dictionary<string, string> keyValuePairs)
+        {
+            // TODO: consider making sure this only fetches once even if the server has high load
+            if (!_cachedOptions.ContainsKey(language))
+            {
+                var langCode = language switch
+                {
+                    "nb" => "1044",
+                    "nn" => "2068",
+                    "en" => "1033",
+                    _ => "1044", // default to norwegian bokmål
+                };
+
+                // Don't use httpClientFactory, as this will only run once per language per container reboot
```

Comment:
Not sure I understood why you can't have the HTTP client injected in the constructor?

Reply:
It's not that I can't. It is just that i thought it would depend on the app being configured with `services.AddHttp()`, and a dependency on `Microsoft.Extensions.Http`, that is currently not in the template. The benefit of using httpClientFactory is to reuse the clients for a limited time, and that is not an issue when it only runs once per language on startup.

Maybe I was wrong? Thinking more about it we probably already use httpClientFactory in the nuget packages, so that was not a valid reason.

Another reason I just came to think of is that I've seen lots of warnings against `static HttpClient = new HttpClient()` because a long lived client might use stale dns data. As this class is registrerd with `AddSingleton`, the same issue would be apparent here, especially because it might be weeks between each time someone opens an app in english or nynorsk.

Reply:
The solution that would work is to inject `IHttpClientFactory` and ask for a client the 3 times it is needed. Seems much more complicated to get the correct lifetime management for such a solution, but maybe I just over complicate things.

Reply:
```suggestion
                // Altinn2CodeListProvider is registrerd with AddSingleton, so injecting a HttpClient would lead to the same issues as 
                // static HttpClient = new HttpClient();
                // This code is cached and only runs once per language per container lifetime,
                // so reuse of the client is not required and we just create a new one each time.
```

Reply:
My main concern is this statement from Microsoft:

> Though this class implements IDisposable, declaring and instantiating it within a using statement is not preferred because when the HttpClient object gets disposed of, the underlying socket is not immediately released, which can lead to a socket exhaustion problem.

From this:
https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests

Seems like the recommended practice is to either let the IHttpClientFactory create the HttpClient or let the DI container inject one.

Reply:
I think I would split the HttpClient part of the code into it's own class and register that as a typed HttpClient. That would put it along side the other typed HttpClients we have eg. ProfileClient and TextClient. You can then inject a IMemoryCache into the typed HttpClient to handle the caching (or decorate it). The typed HttpClient should then be registered with a transient lifetime instead of a singleton. That would leave the options implementation as a thin layer only handling the logic related to building the options list and not http calls or caching.

Reply:
Hmm... I think you'd still get the dns issue with a static HttpClient because options is injected into `AppResourcesSI` that configured using `services.AddSingleton<IAppResources, AppResourcesSI>();`

Reply:
Hm, good point - meaning that the IOptionsService would effectively end up being a singleton as well. And the way AppResourcesSI is implemented it has some simple caching of the Application preventing it from being read from disk every time (which is a good thing) - but I guess that's why it's registered as a singleton. In my opinion this should be changed to take a memory cache in as a dependency, then AppResourcesSI could be transient. I'm not the biggest fan of singleton's - especially in the combination of transient services like this:)

So how can we start to untangle this and use the DI container effectively? I still think creating a separate type http client for talking to Altinn 2 api is a good thing. Then we can keep that logic separate and have it as a transient. We could then take the IHttpClientFactory in as a dependency and ask for a HttpClient. Wouldn't we then get around the issue?

Reply:
Singletons are great, until they need a dependency that can't be singleton.

A complicating factor that the DI container obviously does not clean up Singleton services, but also not Transient services that are referenced by a Singleton. This means that if you need to handle lifetime explicitly, an not just wait for the scope to be torn down at the end of the request, if you don't want to leak memory.

I pushed a simple fix to use `IHttpClientFactory` to get a client. Have a look.

Reply:
Yes, agree - Singleton is a great tool when used correctly. But it is so easy to forget and put a dependency on a transient or scoped service that I think at least twice before I create a singleton. When it comes to option providers I see that they easily can be registered as singletons because of their nature and to get some "free" caching. For some cases options will actually be dependent on the user, making transient a requirement.

Regarding your change I would still like to see the http part, including the IHttpClientFactory, moved out and into a separate typed HttpClient, and that this client is the injected into the options provider. I know it's not much code and that a separate class might seem unnecessary, but we would like to follow the same pattern for all our http clients rather having some of them built into other classes.

The fact that AppResourcesSI is a Singleton today should be changed. It worked well when all resources was static, but with dynamic resources and user specific resources it's a bit to easy to end up returning wrong data. I'll change it and resolve the caching issues as well so that it can be registered as a transient.

Reply:
I think I have changed the code as you requested now.

Looking once more at this, I don't think the problem is that `AppResourcesSI` is a Singleton (though i'm not opposed to changing that). The issue is that `AppResourcesSi` should only load the contents of the files on disk, and that how it was previously working (dynamic was from `App : AppBase`). The dynamic options was not part of `AppResourcesSI` before, and should not be part of it now either.

Reply:
I agree with your observation @ivarne. Ronny and I visited the same reasoning when we talked about this earlier today. 

AppResourcesSI should ideally never have existed. "Handling files" is too broad of a responsibility for one service, but things like that isn't always clear immediatly. Now that the issue has been identified we'll consider splitting it across more specialized services.

Reply:
I didn't think "handling standard files" was too much. Even parsing json/xml and retuning strongly typed objects is fine. The problem arrived when it got the additional responsibility to merge the options from files with options from elsewhere. That was recently introduced when options was changed.

Reply:
Thx for the change @ivarne. I'll start the process of moving it through the pipeline and create a new Nuget if we don't need to support multiple environments. I see more stuff I would like to change, but that's not on you, it's more on the long term of splitting up the AppResourcesSI @SandGrainOne is mentioning - so that is for later.

Reply:
FYI @ivarne picking up this now. I need to manually copy the files to the new repo so I'll close this PR once I'm done and re-create it in the new repo and just merge it and create new packages from there.

Reply:
Great! Just tag me in the new issue so I'll see it.